client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,343 @@
+  ourPlayerIndex: number,
+  deck: CardState[],
+  oldDeck: readonly CardState[],
+) {
+  const cardPossibilitiesForPlayer = getCardPossibilitiesForPlayer(
+    card,
+    playerIndex,
+    ourPlayerIndex,
+  );
+
+  if (
+    card.revealedToPlayer[playerIndex] ||
+    cardPossibilitiesForPlayer.length === 1
+  ) {
+    // The player already knows what this card is.
+    return false;
+  }
+
+  const oldCard = oldDeck[card.order];
+
+  if (typeof oldCard === "undefined" || oldCard.location === "deck") {
+    // this is a newly drawn card and hasn't had any calculations yet.
+    return true;
+  }
+
+  // If the possibilities on the unknown cards don't change, then the result of our calculation
+  // won't change.  We only need to recalculate the card if the input (possibilities) changed.
+  return unknownCardPossibilitiesDifferent(
+    card.order,
+    deck,
+    oldDeck,
+    playerIndex,
+    ourPlayerIndex,
+  );
+}
+
+function getCardPossibilitiesForPlayer(
+  card: CardState,
+  playerIndex: number,
+  ourPlayerIndex: number,
+): ReadonlyArray<readonly [number, number]> {
+  if (card.location === playerIndex) {
+    // If this card is in the players hand, then use our best (empathy) guess.
+    return card.possibleCardsForEmpathy;
+  }
+  if (playerIndex === ourPlayerIndex || card.revealedToPlayer[playerIndex]) {
+    // This is revealed to the player or we are the requested player => just use our best knowledge.
+    return card.possibleCards;
+  }
+  // This is an unrevealed card not in the players hand but not revealed to them.
+  // That can happen with something like a detrimental character (such as 'Slow-Witted')
+  // or throw it in the hole.  We can't use our best (empathy) guess, because it might be in our own
+  // hand and we might know more about the card then the other player does.  We know the other
+  // player at least knows about the clues for it, so we'll use that set of possibilities.
+  return card.possibleCardsFromClues;
+}
+
+function generatePossibilitiesForUnknownCards(
+  excludeCardOrder: number,
+  deck: readonly CardState[],
+  playerIndex: number,
+  ourPlayerIndex: number,
+): ReadonlyArray<ReadonlyArray<readonly [number, number]>> {
+  const unknownCards = generateUnknownCards(
+    excludeCardOrder,
+    deck,
+    playerIndex,
+  );
+  const possibilities: Array<ReadonlyArray<readonly [number, number]>> = [];
+  for (const card of unknownCards) {
+    possibilities.push(
+      getCardPossibilitiesForPlayer(card, playerIndex, ourPlayerIndex),
+    );
+  }
+  // start with the more stable possibilities
+  possibilities.sort((a, b) => a.length - b.length);
+  return possibilities;
+}
+
+function generateUnknownCards(
+  excludeCardOrder: number,
+  deck: readonly CardState[],
+  playerIndex: number,
+) {
+  const unknownCards = [];
+  for (const card of deck) {
+    if (card.order === excludeCardOrder) {
+      continue;
+    }
+    if (card.revealedToPlayer[playerIndex] || card.hasClueApplied) {
```

Comment:
Does this mean that cards in a player's own hand don't get brought in for deduction?

Reply:
There's two sets of cards. I'm bad at naming, but in the code I called them `cardsToCalculate` and `unknownCards`. The `cardsToCalculate` are the cards that we are trying solve the latest `possibleCards` and `possibleCardsForEmpathy` values for. The `unknownCards` are the cards that we're using to disprove as many possibilities (`possibleCards` and/or `possibleCardsForEmpathy`) of each card in `cardsToCalculate`.

A card in the player's own hand might not be added to `unknownCards` because it has no positive or negative clues on it. It would unnecessarily expand the search tree of the `possibilityValid` method because it provides no constraint against the possibility we're checking on the card we're calculating (the unrevealed/unclued card can practically be anything... it would never disprove something). So this improves performance by trimming that search tree.

The `card.order === excludeCardOrder` check is making sure the card we're currently calculating for is not one of the `unknownCards`. Otherwise, we'd be double counting it in the `possibilityValid` method.

If that makes sense, I can try to sprinkle some comments around the code to this effect (or just take those two paragraphs and put it above the highlighted code from the `generateUnknownCards` method).

Reply:
Makes sense. Does that mean that a card that is revealed is considered an unknown card rather than just subtracted from the card count map beforehand?

Reply:
Since I got rid of the `unknownCards` list, in favor if directly iterating, I changed the higher level name/list to `deckPossibilities` in hopes of making more clear... not sure if I accomplished that.

Anyway, to answer your question: Yes, a card that is revealed is still considered an 'unknown card' / 'deckPossibility'.  When I attempted to add code that would subtract it from the `cardCountMap` beforehand it didn't seem to make a noticeable impact, although it's possible I had a bad implementation of that.  Changes to the `cardCountMap` need to be undone by the time we exit the reducer, since it's mutating the cached singleton map.